File: Cracking the Coding Interview Challenges/Queues- A Tale of Two Stacks/soln.py
# Queue as 2 stacks:
    # old to new stack : contains items as they come in
    # new to old stack : contains items in reverse order
class MyQueue(object):
    def __init__(self):
        self.old_new = []
        self.new_old = []

    # For dequeueing, need to pop iteratively from old_new and append to new_old
    # if new_old is not empty, top of new_old is the oldest item - so directly read or pop that value
    # if new_old is empty, pop and append
    def peek(self):
        if not self.new_old:
            while self.old_new:
                self.new_old.append(self.old_new.pop())
        value = self.new_old.pop()
        self.new_old.append(value)
        return value

    def pop(self):
        if not self.new_old:
            while self.old_new:
                self.new_old.append(self.old_new.pop())
        value = self.new_old.pop()
        return value

    def put(self, value):
        # add to old_new stack
        self.old_new.append(value)

queue = MyQueue()
t = int(input())
for line in range(t):
    values = map(int, input().split())
    values = list(values)
    if values[0] == 1:
        queue.put(values[1])
    elif values[0] == 2:
        queue.pop()
    else:
        print(queue.peek())

File: Cracking the Coding Interview Challenges/Queues- A Tale of Two Stacks/soln.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <stack>
#include <queue>
using namespace std;

class MyQueue {
  
    public:
    
        stack<int> stack_newest_on_top, stack_oldest_on_top;   
        void push(int x) {
       
           //First element -> can directly push to the stack_oldest_on_top to save 1 move op later.
            if(stack_newest_on_top.empty()&& stack_oldest_on_top.empty()) 
            {
                stack_oldest_on_top.push(x);
            }
            // All other cases we push to the stack_newest_on_top stack
            else{
                stack_newest_on_top.push(x);
            }
        }

        void pop() {
            
            if(!stack_oldest_on_top.empty())
            {
              // We need to remove the oldest element first. This is why keeping stack_oldest_on_top is useful.
              stack_oldest_on_top.pop();
            }
            // When the stack that stores the oldest one becomes empty, we need to move all the elements from the other 
            // At the end of this op, the order of the queue remains stable while maintaining the same property of the two stacks.
            if(stack_oldest_on_top.empty())
            {
                while(!stack_newest_on_top.empty())
                {
                    int x = stack_newest_on_top.top();
                    stack_newest_on_top.pop();
                    stack_oldest_on_top.push(x);
                }
            }
        }
        // simply read the top of the stack that stores the oldest element on top... 
        int front() {
            return stack_oldest_on_top.top();
        }
};

int main() {
    MyQueue q1;
    int q, type, x;
    cin >> q;
    
    for(int i = 0; i < q; i++) {
        cin >> type;
        if(type == 1) {
            cin >> x;
            q1.push(x);
        }
        else if(type == 2) {
            q1.pop();
        }
        else cout << q1.front() << endl;
    }
    return 0;
}
